### PR TITLE
RideShare Pair - Jane Edwards-McNears/Amanda Ungco

### PR DESCRIPTION
**Pull request by JaneEdwMcN**
# OO Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | We separated the method for request trip - we created a helper method - find available driver. We were considering putting the logic into a big method but chose to separate it in order to make the logic more readable and adaptable.
Describe any examples of composition that you encountered in this project, if any | The compositions we used in this project were one to many for all trip, passenger and driver classes.
Describe the relationship between `User` and `Driver`| Driver is a subclass of User. All drivers are users. The driver class also relates to the user class on a one to many composition.
Describe a _nominal test_ that you wrote for this assignment. | Checking that the total expenditures rounded two places and was a float. 
Describe an _edge case test_ that you wrote for this assignment | Driver can't drive itself. 
Describe a concept that you/your pair gained more clarity on as you worked on this assignment | How classes work and their interconnectedness and calling methods across all classes and files. 
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | We worked well together because we had very open and direct communication with each other and took the time to slowdown and let each other follow their ideas. We used the ping pong method to evenly delegate navigator and driver responsibilities. This strategy also helped us stay on track with writing tests first and pushing to git hub.

### CODE REVIEW

File: lib/user.rb [hunk 1]
```diff
@@ -16,5 +16,100 @@ def initialize(input)
...
+      completed_trips = @trips.select{|trip| trip.end_time != nil}
+      total = completed_trips.reduce(0) { |sum, trip| sum + trip.cost}
+
+    return total
+    end
   end
+
+
+  class Driver < User
+    attr_reader :vehicle_id, :driven_trips, :id, :name, :phone_number, :trips, :driven_trips
```
Comment (dHelmgren) [inline lib/user.rb:40]: You have an `attr_reader` in `User` that has already created `id`, `phone_number` `trips` and `name`, you don't need them redefined in the child class.

Also, `driven_trips` is in your list 2 times, remember to look at the warnings that show up before your test results.

File: specs/trip_spec.rb [hunk 2]
```diff
@@ -9,36 +9,52 @@
...
       end
+
+      it "raises error if end time is before start time" do
+
+        test_trip_input = {:id => 1234, :passenger => "George Foreman", :start_time => "2018-05-25 12:25:00 -0700", :end_time => "2018-05-25 11:52:40 -0700", :cost => 100, :rating => 3}
+
+        expect{ RideShare::Trip.new(test_trip_input) }.must_raise ArgumentError
+      end
+
+      it "calculates ride duration in seconds" do
```
Comment (dHelmgren) [inline specs/trip_spec.rb:52]: What happens if if the trip is incomplete?

File: lib/user.rb [hunk 3]
```diff
@@ -16,5 +16,100 @@ def initialize(input)
...
+    def net_expenditures
+      completed_trips = @trips.select{|trip| trip.end_time != nil}
+      total = completed_trips.reduce(0) { |sum, trip| sum + trip.cost}
+
+    return total
+    end
   end
+
+
+  class Driver < User
```
Comment (dHelmgren) [inline lib/user.rb:39]: Given where you are currently in your coding careers, I would avoid declaring a child class inside the same file as its parent. It can mask bad behavior, leaving your code with sneaky bugs.

File: lib/user.rb [hunk 4]
```diff
@@ -16,5 +16,100 @@ def initialize(input)
...
+
+    def average_rating
+      #exclude costs that == nil
+      if self.driven_trips.length < 1
+        return average_rating = 0
+      end
+
+      ratings_without_nil = self.driven_trips.select{|trip| trip.rating != nil}
+
+      sum_ratings = ratings_without_nil.reduce(0){|memo, trip| memo + trip.rating}
```
Comment (dHelmgren) [inline lib/user.rb:84]: `sum` is probably the clearer choice here. :)

File: specs/driver_spec.rb [hunk 5]
```diff
@@ -1,104 +1,158 @@
...
+      it "returns a float within" do
+        total_revenue = @driver.total_revenue
+        expect(total_revenue).must_be_kind_of Float
+      end
 
-      expect(@driver.average_rating).must_be_close_to (5.0 + 1.0) / 2.0, 0.01
+      it "returns zero if no driven trips" do
+        driver_2 = RideShare::Driver.new(id: 54, name: "Rogers Bartell IV", vin: "1C9EVBRM0YBC564DZ")
+
+        expect(driver_2.total_revenue).must_equal 0
```
Comment (dHelmgren) [inline specs/driver_spec.rb:125]: What if you have an incomplete trip?

File: specs/trip_dispatch_spec.rb [hunk 6]
```diff
@@ -111,4 +108,90 @@
...
+      test_trip_input = {:id => 1234, :passenger => "George Foreman", :end_time => Time.parse("2018-05-25 12:25:00 -0700"), :start_time => Time.parse("2018-05-25 11:52:40 -0700"), :cost => 100, :rating => 3}
+
+      test_ride = RideShare::Trip.new(test_trip_input)
+
+
+      @dispatcher.drivers[0].status = :AVAILABLE
+      @dispatcher.drivers[2].add_driven_trip(test_ride)
+
+      expect(@dispatcher.request_trip(1).driver.id).must_equal 7
+
```
Comment (dHelmgren) [inline specs/trip_dispatch_spec.rb:192]: What if there is no available driver? This is an important case, and is exactly the sort of situation where software will do something unexpected. You might set up the test as follows:

```ruby
it 'raises an error if no driver is available' do
  @dispatcher.drivers.each do |driver|
    # This may require an attr_writer in Driver
    driver.status = :UNAVAILABLE
  end
  proc {
    @dispatcher.request_trip(2)
  }.must_raise StandardError
end
```

File: lib/trip_dispatcher.rb [hunk 7]
```diff
@@ -60,11 +105,62 @@ def find_passenger(id)
...
+
+      available_drivers = available_drivers.reject{|driver| driver.id == user_id}
+
+      if available_drivers.length == 0
+        raise ArgumentError, "There are no available drivers"
+      else
+        available_drivers.each do |driver|
+          if driver.driven_trips.length == 0
+            driver.status = :UNAVAILABLE
+            return driver
```
Comment (dHelmgren) [inline lib/trip_dispatcher.rb:154]: Nice! I love early returns!

**Comment by dHelmgren** [comment]
# Ride Share
## What We're Looking For

Feature|Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | Great job!
Answer comprehension questions | Good answers! 
**Wave 1** |
Appropriate use of Ruby's `Date`| Looks good!
`Trip` has a helper method to calculate duration| check!
`User` (passenger) has a method to calculate total cost of all trips| Yep!
Tests for wave 1 | Good tests!
**Wave 2** |
`Driver` inherits from `User` | Yes!
`Driver` has `add_driven_trip` method | yeo!
`Driver` has method to calculate average rating | Yep!
`Driver` has method to calculate net expenditures and it uses `super` | So, your method doesn't use `super`. Since you have a method in `User` with the same name, instead of calculating the expenditures again, you could have invoked the parent's version directly just by using the keyword `super`. If you have more questions about this, lets talk!
`Driver` has a method to calculate total revenue | sure does!
Tests for wave 2 | Tests for wave 2 are good!
**Wave 3** |
`TripDispatcher` has a new method to create trips | Yep!
creating a trip in `TripDispatcher` relies on methods in `Driver` and `User` (passenger) to modify their own attributes | Yes!
Complex logic was correctly implemented | Yes!
Tests for `request_trip` | You're handling of nominal cases was good, but you missed an important edge case. 
Methods from wave 1 and 2 handle incomplete trips | They handle incomplete trips...
Tests for wave 1 and 2 methods with incomplete trips | But you didn't test for them in `Driver` or `User`!
**Wave 4 (Optional)** |
`TripDispatcher` now assigns trips to either the newest driver (no trips), or the driver who has not driven in the longest time | Looks good.
Appropriate helper methods were made to help with complex logic | Yeah! this was SUPER EASY TO READ!
Tests for wave 4 | I didn't spot any tests that checked your logic.
**Overall** | Your submission was solid in the code, but I think you need to make sure that you are thorough about testing! Its easy to make a big change to the way your application handles data late in the process and not properly test how that affects other areas of code (incomplete trips being the example from this assignment).

File: specs/trip_dispatch_spec.rb (hunk 6 above)
Comment (unkcodesquick) [reply specs/trip_dispatch_spec.rb:192]: dang! we put an argumenterror in trip_dispatcher.rb line 148 but must've forgotten to write a test.
